### PR TITLE
Chapter 02 SXG

### PR DESCRIPTION
**Pull request by sebing**

### CODE REVIEW

File: chapter02/sxg/src/main/java/com/github/sebin/chapter02/AccountTransaction.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package com.github.sebin.chapter02;
+
+import java.util.Calendar;
+
+public class AccountTransaction {
+	private int fromAccountId;
+	private int toAccountId;
+	private int amount;
+	private Calendar time;
```
Comment (qianliok) [inline chapter02/sxg/src/main/java/com/github/sebin/chapter02/AccountTransaction.java]: Nice, I totally forgot about the timestamp, Yes I believe we do need a timestamp for the transaction.

File: chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	private String fullname;
+	
+	/**
+	 * @param customerID
+	 * @param username
+	 * @param password
+	 * @param fullname
+	 */
+	public User(int customerID, String username, String password, String fullname) {
+		super();
```
Comment (qianliok) [inline chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java]: super()? maybe not needed

File: chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package com.github.sebin.chapter02;
+
+public class User {
+	private int customerID;
+	private String username;
+	private String password;
+	private String fullname;
```
Comment (qianliok) [inline chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java]: there is no setters, so these parameters are `final`, back to the old topic, are you sure customer's name will never change at all? :smile:

Comment (sebing) [inline chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java]: When the customer changes his name, create a new User object. Let the old one get garbage collected.

I want to have classes as much as possible to be immutable.

File: chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java (hunk 2 above)
Comment (sebing) [inline chapter02/sxg/src/main/java/com/github/sebin/chapter02/User.java]: may be needed when we decided to change the super class from Object to something else..:)

File: chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		}
+	}
+
+	@Test(expected=IllegalStateException.class)
+	public void canLimitCreditAmount() {
+		ApplicationContext context = new ClassPathXmlApplicationContext("context.xml");
+		BankService bank = (BankService)context.getBean("bank");
+		Account account = (Account)context.getBean("account");
+		double amount = account.getAccountMaxBalance()+1;
+		bank.credit(account, amount);
```
Comment (qianliok) [inline chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java]: What are we checking in this test method? don't see any assertions

Comment (sebing) [inline chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java]: haha... my mistake...It is expecting a wrong transaction(expecting an exception)

File: chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		ApplicationContext context = new ClassPathXmlApplicationContext("context.xml");
+		BankService bank = (BankService)context.getBean("bank");
+		Account account = (Account)context.getBean("account");
+		double initValue = account.getAccountBalance();
+		double amount = 5;
+		if( (initValue-amount)>=0){
+			bank.debit(account, amount);
+			Assert.assertEquals(account.getAccountBalance(),(initValue-amount),0);
+		}else{
+			Assert.assertTrue(false); // can't test, there is not enough money.
```
Comment (qianliok) [inline chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java]: I think you can test this if you set the account with a init balance less than the amount, or maybe raise the amount more than your default init balance of 10?

Comment (sebing) [inline chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java]: hihi.. you are right.. I forgot account had a set method.

File: chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java (hunk 4 above)
Comment (qianliok) [inline chapter02/sxg/src/test/java/com/github/sebin/chapter02/BankServiceTest.java]: oh cool, I see, there is Test expected Exception annotation.

**Pull request merged**